File: sdk/python/src/delos/services/eval.py
"""Client for the eval service."""

from __future__ import annotations

import sys
from datetime import timezone
from typing import TYPE_CHECKING

from delos.models.eval import (
    EvalConfig,
    EvalResult,
    EvalRun,
    EvalRunStatus,
    EvalSummary,
    Evaluator,
    EvaluatorConfig,
    EvaluatorParam,
    EvaluatorResult,
    ExampleComparison,
    RunComparison,
)
from delos.services.base import BaseClient

if TYPE_CHECKING:
    from delos.config import ServiceEndpoint

sys.path.insert(0, "gen/python")

try:
    from eval.v1 import eval_pb2, eval_pb2_grpc
except ImportError:
    eval_pb2 = None  # type: ignore
    eval_pb2_grpc = None  # type: ignore


class EvalClient(BaseClient):
    """Client for the eval service."""

    def __init__(self, endpoint: ServiceEndpoint, timeout: float = 30.0) -> None:
        """Initialize the eval client."""
        super().__init__(endpoint, timeout)
        self._stub: eval_pb2_grpc.EvalServiceStub | None = None

    @property
    def stub(self) -> eval_pb2_grpc.EvalServiceStub:
        """Get the gRPC stub."""
        if self._stub is None:
            if eval_pb2_grpc is None:
                raise ImportError("Generated protobuf code not found. Run 'buf generate' first.")
            self._stub = eval_pb2_grpc.EvalServiceStub(self.channel)
        return self._stub

    def create_run(
        self,
        name: str,
        *,
        description: str = "",
        prompt_id: str = "",
        prompt_version: int = 0,
        dataset_id: str = "",
        config: EvalConfig | None = None,
        metadata: dict[str, str] | None = None,
    ) -> EvalRun:
        """Create and start an evaluation run."""
        pb_config = None
        if config:
            pb_config = eval_pb2.EvalConfig(
                evaluators=[
                    eval_pb2.EvaluatorConfig(
                        type=e.type,
                        name=e.name,
                        params=e.params,
                        weight=e.weight,
                    )
                    for e in config.evaluators
                ],
                provider=config.provider,
                model=config.model,
                concurrency=config.concurrency,
                sample_size=config.sample_size,
                shuffle=config.shuffle,
            )

        request = eval_pb2.CreateEvalRunRequest(
            name=name,
            description=description,
            prompt_id=prompt_id,
            prompt_version=prompt_version,
            dataset_id=dataset_id,
            config=pb_config,
            metadata=metadata or {},
        )

        response = self.stub.CreateEvalRun(request, timeout=self._timeout)
        return self._to_run(response.eval_run)

    def get_run(self, id: str) -> EvalRun | None:
        """Get an evaluation run by ID."""
        request = eval_pb2.GetEvalRunRequest(id=id)
        try:
            response = self.stub.GetEvalRun(request, timeout=self._timeout)
            return self._to_run(response.eval_run)
        except Exception:
            return None

    def list_runs(
        self,
        *,
        prompt_id: str = "",
        dataset_id: str = "",
        status: EvalRunStatus = EvalRunStatus.UNSPECIFIED,
        limit: int = 100,
        offset: int = 0,
    ) -> tuple[list[EvalRun], int]:
        """List evaluation runs."""
        request = eval_pb2.ListEvalRunsRequest(
            prompt_id=prompt_id,
            dataset_id=dataset_id,
            status=self._status_to_pb(status),
            limit=limit,
            offset=offset,
        )
        response = self.stub.ListEvalRuns(request, timeout=self._timeout)
        runs = [self._to_run(r) for r in response.eval_runs]
        return runs, response.total_count

    def cancel_run(self, id: str) -> EvalRun:
        """Cancel a running evaluation."""
        request = eval_pb2.CancelEvalRunRequest(id=id)
        response = self.stub.CancelEvalRun(request, timeout=self._timeout)
        return self._to_run(response.eval_run)

    def get_results(
        self,
        eval_run_id: str,
        *,
        failed_only: bool = False,
        limit: int = 100,
        offset: int = 0,
    ) -> tuple[list[EvalResult], int]:
        """Get results for an evaluation run."""
        request = eval_pb2.GetEvalResultsRequest(
            eval_run_id=eval_run_id,
            failed_only=failed_only,
            limit=limit,
            offset=offset,
        )
        response = self.stub.GetEvalResults(request, timeout=self._timeout)
        results = [self._to_result(r) for r in response.results]
        return results, response.total_count

    def compare_runs(
        self,
        run_id_a: str,
        run_id_b: str,
    ) -> tuple[RunComparison, RunComparison, list[ExampleComparison]]:
        """Compare two evaluation runs."""
        request = eval_pb2.CompareRunsRequest(
            run_id_a=run_id_a,
            run_id_b=run_id_b,
        )
        response = self.stub.CompareRuns(request, timeout=self._timeout)

        run_a = RunComparison(
            run_id=response.run_a.run_id,
            prompt_version=response.run_a.prompt_version,
            overall_score=response.run_a.overall_score,
            pass_rate=response.run_a.pass_rate,
            avg_latency_ms=response.run_a.avg_latency_ms,
            total_cost_usd=response.run_a.total_cost_usd,
        )
        run_b = RunComparison(
            run_id=response.run_b.run_id,
            prompt_version=response.run_b.prompt_version,
            overall_score=response.run_b.overall_score,
            pass_rate=response.run_b.pass_rate,
            avg_latency_ms=response.run_b.avg_latency_ms,
            total_cost_usd=response.run_b.total_cost_usd,
        )
        examples = [
            ExampleComparison(
                example_id=e.example_id,
                score_a=e.score_a,
                score_b=e.score_b,
                score_diff=e.score_diff,
                regression=e.regression,
            )
            for e in response.examples
        ]
        return run_a, run_b, examples

    def list_evaluators(self) -> list[Evaluator]:
        """List available evaluator types."""
        request = eval_pb2.ListEvaluatorsRequest()
        response = self.stub.ListEvaluators(request, timeout=self._timeout)
        return [
            Evaluator(
                type=e.type,
                name=e.name,
                description=e.description,
                params=[
                    EvaluatorParam(
                        name=p.name,
                        type=p.type,
                        description=p.description,
                        required=p.required,
                        default_value=p.default_value,
                    )
                    for p in e.params
                ],
            )
            for e in response.evaluators
        ]

    def _to_run(self, pb: eval_pb2.EvalRun) -> EvalRun:
        """Convert protobuf to model."""
        config = None
        if pb.HasField("config"):
            config = EvalConfig(
                evaluators=[
                    EvaluatorConfig(
                        type=e.type,
                        name=e.name,
                        params=dict(e.params),
                        weight=e.weight,
                    )
                    for e in pb.config.evaluators
                ],
                provider=pb.config.provider,
                model=pb.config.model,
                concurrency=pb.config.concurrency,
                sample_size=pb.config.sample_size,
                shuffle=pb.config.shuffle,
            )

        summary = None
        if pb.HasField("summary"):
            summary = EvalSummary(
                overall_score=pb.summary.overall_score,
                scores_by_evaluator=dict(pb.summary.scores_by_evaluator),
                passed_count=pb.summary.passed_count,
                failed_count=pb.summary.failed_count,
                pass_rate=pb.summary.pass_rate,
                total_cost_usd=pb.summary.total_cost_usd,
                total_tokens=pb.summary.total_tokens,
                avg_latency_ms=pb.summary.avg_latency_ms,
            )

        return EvalRun(
            id=pb.id,
            name=pb.name,
            description=pb.description,
            prompt_id=pb.prompt_id,
            prompt_version=pb.prompt_version,
            dataset_id=pb.dataset_id,
            config=config,
            status=self._status_from_pb(pb.status),
            error_message=pb.error_message,
            total_examples=pb.total_examples,
            completed_examples=pb.completed_examples,
            summary=summary,
            created_at=pb.created_at.ToDatetime(timezone.utc) if pb.HasField("created_at") else None,
            started_at=pb.started_at.ToDatetime(timezone.utc) if pb.HasField("started_at") else None,
            completed_at=pb.completed_at.ToDatetime(timezone.utc) if pb.HasField("completed_at") else None,
            created_by=pb.created_by,
            metadata=dict(pb.metadata),
        )

    def _to_result(self, pb: eval_pb2.EvalResult) -> EvalResult:
        """Convert protobuf result to model."""
        from google.protobuf.json_format import MessageToDict

        evaluator_results = {}
        for k, v in pb.evaluator_results.items():
            evaluator_results[k] = EvaluatorResult(
                evaluator_type=v.evaluator_type,
                score=v.score,
                passed=v.passed,
                explanation=v.explanation,
                details=dict(v.details),
            )

        return EvalResult(
            id=pb.id,
            eval_run_id=pb.eval_run_id,
            example_id=pb.example_id,
            input=MessageToDict(pb.input) if pb.HasField("input") else {},
            expected_output=MessageToDict(pb.expected_output) if pb.HasField("expected_output") else {},
            actual_output=MessageToDict(pb.actual_output) if pb.HasField("actual_output") else {},
            evaluator_results=evaluator_results,
            overall_score=pb.overall_score,
            passed=pb.passed,
            latency_ms=pb.latency_ms,
            tokens_used=pb.tokens_used,
            cost_usd=pb.cost_usd,
            error=pb.error,
        )

    def _status_to_pb(self, status: EvalRunStatus) -> int:
        """Convert status to protobuf."""
        mapping = {
            EvalRunStatus.UNSPECIFIED: eval_pb2.EVAL_RUN_STATUS_UNSPECIFIED,
            EvalRunStatus.PENDING: eval_pb2.EVAL_RUN_STATUS_PENDING,
            EvalRunStatus.RUNNING: eval_pb2.EVAL_RUN_STATUS_RUNNING,
            EvalRunStatus.COMPLETED: eval_pb2.EVAL_RUN_STATUS_COMPLETED,
            EvalRunStatus.FAILED: eval_pb2.EVAL_RUN_STATUS_FAILED,
            EvalRunStatus.CANCELLED: eval_pb2.EVAL_RUN_STATUS_CANCELLED,
        }
        return mapping.get(status, eval_pb2.EVAL_RUN_STATUS_UNSPECIFIED)

    def _status_from_pb(self, status: int) -> EvalRunStatus:
        """Convert protobuf to status."""
        mapping = {
            eval_pb2.EVAL_RUN_STATUS_UNSPECIFIED: EvalRunStatus.UNSPECIFIED,
            eval_pb2.EVAL_RUN_STATUS_PENDING: EvalRunStatus.PENDING,
            eval_pb2.EVAL_RUN_STATUS_RUNNING: EvalRunStatus.RUNNING,
            eval_pb2.EVAL_RUN_STATUS_COMPLETED: EvalRunStatus.COMPLETED,
            eval_pb2.EVAL_RUN_STATUS_FAILED: EvalRunStatus.FAILED,
            eval_pb2.EVAL_RUN_STATUS_CANCELLED: EvalRunStatus.CANCELLED,
        }
        return mapping.get(status, EvalRunStatus.UNSPECIFIED)

File: cli/cmd/eval.go
package cmd

import (
	"context"
	"fmt"
	"time"

	"github.com/spf13/cobra"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"

	evalv1 "github.com/instantcocoa/delos/gen/go/eval/v1"
	"github.com/instantcocoa/delos/cli/internal/output"
)

var evalCmd = &cobra.Command{
	Use:   "eval",
	Short: "Evaluation operations",
	Long:  "Commands for running and managing evaluations.",
}

var evalRunCmd = &cobra.Command{
	Use:   "run",
	Short: "Create and start an evaluation run",
	RunE: func(cmd *cobra.Command, args []string) error {
		conn, err := grpc.NewClient(cfg.EvalAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
		if err != nil {
			return fmt.Errorf("failed to connect: %w", err)
		}
		defer conn.Close()

		client := evalv1.NewEvalServiceClient(conn)
		ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
		defer cancel()

		name, _ := cmd.Flags().GetString("name")
		promptID, _ := cmd.Flags().GetString("prompt")
		promptVersion, _ := cmd.Flags().GetInt32("version")
		datasetID, _ := cmd.Flags().GetString("dataset")
		evaluators, _ := cmd.Flags().GetStringSlice("evaluators")

		if name == "" {
			name = fmt.Sprintf("eval-%s", time.Now().Format("20060102-150405"))
		}

		evalConfigs := make([]*evalv1.EvaluatorConfig, len(evaluators))
		for i, e := range evaluators {
			evalConfigs[i] = &evalv1.EvaluatorConfig{
				Type:   e,
				Weight: 1.0,
			}
		}

		resp, err := client.CreateEvalRun(ctx, &evalv1.CreateEvalRunRequest{
			Name:          name,
			PromptId:      promptID,
			PromptVersion: promptVersion,
			DatasetId:     datasetID,
			Config: &evalv1.EvalConfig{
				Evaluators:  evalConfigs,
				Concurrency: 5,
			},
		})
		if err != nil {
			return fmt.Errorf("failed to create eval run: %w", err)
		}

		output.Success("Created evaluation run %s (ID: %s)", resp.EvalRun.Name, resp.EvalRun.Id)
		output.Info("Status: %s", resp.EvalRun.Status.String())

		return nil
	},
}

var evalListCmd = &cobra.Command{
	Use:   "list",
	Short: "List evaluation runs",
	RunE: func(cmd *cobra.Command, args []string) error {
		conn, err := grpc.NewClient(cfg.EvalAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
		if err != nil {
			return fmt.Errorf("failed to connect: %w", err)
		}
		defer conn.Close()

		client := evalv1.NewEvalServiceClient(conn)
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		promptID, _ := cmd.Flags().GetString("prompt")
		datasetID, _ := cmd.Flags().GetString("dataset")

		resp, err := client.ListEvalRuns(ctx, &evalv1.ListEvalRunsRequest{
			PromptId:  promptID,
			DatasetId: datasetID,
			Limit:     100,
		})
		if err != nil {
			return fmt.Errorf("failed to list eval runs: %w", err)
		}

		if cfg.Format == "json" || cfg.Format == "yaml" {
			w := output.NewWriter(cfg.Format)
			return w.Print(resp.EvalRuns)
		}

		table := output.Table{
			Headers: []string{"ID", "NAME", "STATUS", "PROGRESS", "SCORE", "CREATED"},
			Rows:    make([][]string, len(resp.EvalRuns)),
		}
		for i, r := range resp.EvalRuns {
			created := ""
			if r.CreatedAt != nil {
				created = r.CreatedAt.AsTime().Format("2006-01-02 15:04")
			}
			progress := fmt.Sprintf("%d/%d", r.CompletedExamples, r.TotalExamples)
			score := "-"
			if r.Summary != nil {
				score = fmt.Sprintf("%.2f", r.Summary.OverallScore)
			}
			table.Rows[i] = []string{
				r.Id[:8],
				r.Name,
				r.Status.String(),
				progress,
				score,
				created,
			}
		}

		w := output.NewWriter("table")
		return w.Print(table)
	},
}

var evalGetCmd = &cobra.Command{
	Use:   "get <id>",
	Short: "Get evaluation run details",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		conn, err := grpc.NewClient(cfg.EvalAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
		if err != nil {
			return fmt.Errorf("failed to connect: %w", err)
		}
		defer conn.Close()

		client := evalv1.NewEvalServiceClient(conn)
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		resp, err := client.GetEvalRun(ctx, &evalv1.GetEvalRunRequest{Id: args[0]})
		if err != nil {
			return fmt.Errorf("failed to get eval run: %w", err)
		}

		w := output.NewWriter(cfg.Format)
		return w.Print(resp.EvalRun)
	},
}

var evalCancelCmd = &cobra.Command{
	Use:   "cancel <id>",
	Short: "Cancel an evaluation run",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		conn, err := grpc.NewClient(cfg.EvalAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
		if err != nil {
			return fmt.Errorf("failed to connect: %w", err)
		}
		defer conn.Close()

		client := evalv1.NewEvalServiceClient(conn)
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		_, err = client.CancelEvalRun(ctx, &evalv1.CancelEvalRunRequest{Id: args[0]})
		if err != nil {
			return fmt.Errorf("failed to cancel eval run: %w", err)
		}

		output.Success("Cancelled evaluation run %s", args[0])
		return nil
	},
}

var evalResultsCmd = &cobra.Command{
	Use:   "results <run-id>",
	Short: "Get evaluation results",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		conn, err := grpc.NewClient(cfg.EvalAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
		if err != nil {
			return fmt.Errorf("failed to connect: %w", err)
		}
		defer conn.Close()

		client := evalv1.NewEvalServiceClient(conn)
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		failedOnly, _ := cmd.Flags().GetBool("failed")
		limit, _ := cmd.Flags().GetInt32("limit")

		resp, err := client.GetEvalResults(ctx, &evalv1.GetEvalResultsRequest{
			EvalRunId:  args[0],
			FailedOnly: failedOnly,
			Limit:      limit,
		})
		if err != nil {
			return fmt.Errorf("failed to get results: %w", err)
		}

		output.Info("Found %d results (showing %d)", resp.TotalCount, len(resp.Results))

		w := output.NewWriter(cfg.Format)
		return w.Print(resp.Results)
	},
}

var evalCompareCmd = &cobra.Command{
	Use:   "compare <run-a> <run-b>",
	Short: "Compare two evaluation runs",
	Args:  cobra.ExactArgs(2),
	RunE: func(cmd *cobra.Command, args []string) error {
		conn, err := grpc.NewClient(cfg.EvalAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
		if err != nil {
			return fmt.Errorf("failed to connect: %w", err)
		}
		defer conn.Close()

		client := evalv1.NewEvalServiceClient(conn)
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		resp, err := client.CompareRuns(ctx, &evalv1.CompareRunsRequest{
			RunIdA: args[0],
			RunIdB: args[1],
		})
		if err != nil {
			return fmt.Errorf("failed to compare runs: %w", err)
		}

		w := output.NewWriter(cfg.Format)
		return w.Print(resp)
	},
}

var evalEvaluatorsCmd = &cobra.Command{
	Use:   "evaluators",
	Short: "List available evaluators",
	RunE: func(cmd *cobra.Command, args []string) error {
		conn, err := grpc.NewClient(cfg.EvalAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
		if err != nil {
			return fmt.Errorf("failed to connect: %w", err)
		}
		defer conn.Close()

		client := evalv1.NewEvalServiceClient(conn)
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		resp, err := client.ListEvaluators(ctx, &evalv1.ListEvaluatorsRequest{})
		if err != nil {
			return fmt.Errorf("failed to list evaluators: %w", err)
		}

		if cfg.Format == "json" || cfg.Format == "yaml" {
			w := output.NewWriter(cfg.Format)
			return w.Print(resp.Evaluators)
		}

		table := output.Table{
			Headers: []string{"TYPE", "NAME", "DESCRIPTION"},
			Rows:    make([][]string, len(resp.Evaluators)),
		}
		for i, e := range resp.Evaluators {
			desc := e.Description
			if len(desc) > 50 {
				desc = desc[:47] + "..."
			}
			table.Rows[i] = []string{e.Type, e.Name, desc}
		}

		w := output.NewWriter("table")
		return w.Print(table)
	},
}

func init() {
	// Run flags
	evalRunCmd.Flags().String("name", "", "Run name")
	evalRunCmd.Flags().String("prompt", "", "Prompt ID")
	evalRunCmd.Flags().Int32("version", 0, "Prompt version (0 = latest)")
	evalRunCmd.Flags().String("dataset", "", "Dataset ID")
	evalRunCmd.Flags().StringSlice("evaluators", []string{"exact_match"}, "Evaluator types")

	// List flags
	evalListCmd.Flags().String("prompt", "", "Filter by prompt ID")
	evalListCmd.Flags().String("dataset", "", "Filter by dataset ID")

	// Results flags
	evalResultsCmd.Flags().Bool("failed", false, "Only show failed results")
	evalResultsCmd.Flags().Int32("limit", 20, "Max results to show")

	evalCmd.AddCommand(evalRunCmd)
	evalCmd.AddCommand(evalListCmd)
	evalCmd.AddCommand(evalGetCmd)
	evalCmd.AddCommand(evalCancelCmd)
	evalCmd.AddCommand(evalResultsCmd)
	evalCmd.AddCommand(evalCompareCmd)
	evalCmd.AddCommand(evalEvaluatorsCmd)
}
